core/src/main/java/org/apache/iceberg/BaseFile.java

```diff
@@ -185,13 +188,30 @@ public PartitionData copy() {
     this.partitionType = toCopy.partitionType;
     this.recordCount = toCopy.recordCount;
     this.fileSizeInBytes = toCopy.fileSizeInBytes;
-    if (fullCopy) {
-      this.columnSizes = SerializableMap.copyOf(toCopy.columnSizes);
-      this.valueCounts = SerializableMap.copyOf(toCopy.valueCounts);
-      this.nullValueCounts = SerializableMap.copyOf(toCopy.nullValueCounts);
-      this.nanValueCounts = SerializableMap.copyOf(toCopy.nanValueCounts);
-      this.lowerBounds = SerializableByteBufferMap.wrap(SerializableMap.copyOf(toCopy.lowerBounds));
-      this.upperBounds = SerializableByteBufferMap.wrap(SerializableMap.copyOf(toCopy.upperBounds));
+    if (copyStats) {
+      if (requestedColumnIds == null) {
```

Comment:
Optional: I'd consider adding a few helper methods to simplify this block. We will have 3 quite big branches now. I understand we will do more null checks but I doubt it will have any noticeable performance impact.

```
private static <K, V> Map<K, V> copyMap(Map<K, V> map, Set<K> keys) {
  return keys == null ? SerializableMap.copyOf(map) : SerializableMap.filteredCopyOf(map, keys);
}

private static Map<Integer, ByteBuffer> copyByteBufferMap(
    Map<Integer, ByteBuffer> map, Set<Integer> keys) {
  return SerializableByteBufferMap.wrap(copyMap(map, keys));
}

...

if (copyStats) {
  this.columnSizes = copyMap(toCopy.columnSizes, requestedColumnIds);
  this.valueCounts = copyMap(toCopy.valueCounts, requestedColumnIds);
  this.nullValueCounts = copyMap(toCopy.nullValueCounts, requestedColumnIds);
  this.nanValueCounts = copyMap(toCopy.nanValueCounts, requestedColumnIds);
  this.lowerBounds = copyByteBufferMap(toCopy.lowerBounds, requestedColumnIds);
  this.upperBounds = copyByteBufferMap(toCopy.upperBounds, requestedColumnIds);
}
```

Up to you here, @pvary. Keep it as is if you want.

Reply:
Applied the solution suggested by you